File: gen_index.py
#!/usr/bin/env python3
from pathlib import Path

root_dir = Path("./wiki")
exclude_dir = {".vitepress", "public"}


def generate_structure():
    lines = []
    # 遍历根目录，跳过非笔记目录
    for category_dir in sorted(
        (
            entry
            for entry in root_dir.iterdir()
            if entry.is_dir() and entry.name not in exclude_dir
        ),
        key=lambda x: x.name,
    ):
        category_name = category_dir.name
        lines.append(f"# {category_name}\n")
        # 遍历子目录，找到所有笔记
        for note_dir in sorted(
            (subdir for subdir in category_dir.iterdir() if subdir.is_dir()),
            key=lambda x: x.name,
        ):
            note_folder = note_dir.name
            indexmd_path = note_dir / "index.md"
            note_title = note_folder
            # 获取每个笔记的标题
            if indexmd_path.exists() and indexmd_path.is_file():
                try:
                    content = indexmd_path.read_text(encoding="utf-8")
                    for line in content.splitlines():
                        line = line.strip()
                        if line.startswith("# "):
                            note_title = line[2:].strip()
                            break
                except Exception as e:
                    print(f"err in {indexmd_path}:{e}")
            lines.append(f"- [{note_title}](/{category_name}/{note_folder}/)\n")

    return "\n".join(lines)


def insert_into_index(content):
    index_path = root_dir / "index.md"
    if not index_path.exists():
        existing_lines = []
    else:
        existing_lines = index_path.read_text(encoding="utf-8").splitlines(
            keepends=True
        )

    keep = existing_lines[:4] if len(existing_lines) >= 4 else existing_lines
    new_lines = keep + [content + "\n"]
    index_path.write_text("".join(new_lines), encoding="utf-8")


if __name__ == "__main__":
    structure = generate_structure()
    insert_into_index(structure)
    print("Update index successfully\n")

File: gen_index.ts
import { promises as fs } from "fs";
import { join } from "path";

const rootDir = "./wiki";
const excludeDir = new Set(["public"]);

// 读取 index.md 的标题
async function getNoteTitle(
    indexPath: string,
    fallback: string,
): Promise<string> {
    try {
        const content = await fs.readFile(indexPath, "utf-8");
        const titleLine = content
            .split("\n")
            .map((l) => l.trim())
            .find((l) => l.startsWith("# "));
        if (titleLine) return titleLine.replace(/^#\s+/, "").trim();
    } catch {
        // ignore
    }
    return fallback;
}

// 整理目录 把笔记前面，分类放后面
async function sortDirsByType(parentPath: string, dirs: any[]): Promise<any[]> {
    const withIndex: any[] = [];
    const withoutIndex: any[] = [];

    for (const dir of dirs) {
        const indexPath = join(parentPath, dir.name, "index.md");
        const hasIndex = await fileExists(indexPath);
        if (hasIndex) {
            withIndex.push(dir);
        } else {
            withoutIndex.push(dir);
        }
    }

    // 各自按字母顺序排列
    withIndex.sort((a, b) => a.name.localeCompare(b.name));
    withoutIndex.sort((a, b) => a.name.localeCompare(b.name));

    return [...withIndex, ...withoutIndex];
}

// 递归遍历目录
async function walkDir(
    dirPath: string,
    relativePath: string,
    indentLevel = 0,
): Promise<string[]> {
    const lines: string[] = [];
    const entries = await fs.readdir(dirPath, { withFileTypes: true });

    // 只看文件夹
    const dirs = entries.filter((e) => e.isDirectory());
    const orderedDirs = await sortDirsByType(dirPath, dirs);

    for (const dir of orderedDirs) {
        const currentPath = join(dirPath, dir.name);
        const indexPath = join(currentPath, "index.md");

        // 判断该目录是否是笔记目录，有 index.md
        const hasIndex =
            entries.some((e) => e.name === dir.name && e.isDirectory()) &&
            (await fs.readdir(currentPath)).includes("index.md");

        // 判断是否还有子目录，可能是分类
        const subEntries = await fs.readdir(currentPath, {
            withFileTypes: true,
        });
        const subDirs = subEntries.filter((e) => e.isDirectory());

        const indent = "    ".repeat(indentLevel);

        if (subDirs.length > 0 && !hasIndex) {
            // 是分类目录（没有 index.md）
            lines.push(`${indent}- **${dir.name}**`);
            const subLines = await walkDir(
                currentPath,
                join(relativePath, dir.name),
                indentLevel + 1,
            );
            lines.push(...subLines);
        } else if (await fileExists(indexPath)) {
            // 是笔记目录（有 index.md）
            const title = await getNoteTitle(indexPath, dir.name);
            lines.push(
                `${indent}- [${title}](/${join(relativePath, dir.name)}/)`,
            );
        }
    }

    return lines;
}

async function fileExists(path: string): Promise<boolean> {
    try {
        await fs.access(path);
        return true;
    } catch {
        return false;
    }
}

async function generateStructure(): Promise<string> {
    const lines: string[] = [];
    const entries = await fs.readdir(rootDir, { withFileTypes: true });

    const categoryDirs = entries
        .filter((e) => e.isDirectory() && !excludeDir.has(e.name))
        .sort((a, b) => a.name.localeCompare(b.name));

    for (const category of categoryDirs) {
        const categoryName = category.name;
        lines.push(`## ${categoryName}\n`);

        const categoryPath = join(rootDir, categoryName);
        const subLines = await walkDir(categoryPath, categoryName, 0);
        lines.push(...subLines, ""); // 空行分隔
    }

    return lines.join("\n");
}

const content = await generateStructure();

//读取 index.md
const indexPath = join(rootDir, "index.md");
let existingLines: string[] = [];
const indexMD = await fs.readFile(indexPath, "utf-8");
existingLines = indexMD.split(/\r?\n/);

// 保留前五行 frontmatter
const keep = existingLines.slice(0, 10);
const newContent = [...keep, "", content, ""].join("\n");
await fs.writeFile(indexPath, newContent, "utf-8");

console.log("Update index successfully\n");
